### DISCUSSION TITLE
<Transition> element exits without animating when content is changed with setState

### DISCUSSION THREAD

**Discussion by jelacs** [pmndrs/react-spring#1327] [Support]
## 🤓 Question
Reposting my question as it has been closed without being solved.
I'm trying to create a list of expandable and collapsible items whose content comes from a single property in component state. The content is supposed to unmount on collapse as it is first removed from state. Animation works perfectly when entering but upon exit animation it just disappears (if expanded item is clicked to collapse) or is updated with the new state (if another item is clicked). 
I could not post actual code but I tried to replicate the issue in [this project](https://codesandbox.io/s/react-spring-state-change-8zztf).

I tried asking on StackOverflow but no luck. I hope someone here could help me find a solution. TIA!

**Comment by harshagarwalnyu** [comment]
This is a classic react-spring gotcha: the component disappears without exit animation because React removes it from the DOM before react-spring can play the `leave` transition.

**Root cause:** you're controlling visibility via state (`if (isOpen)`) outside of react-spring. When state changes, React unmounts immediately.

**Fix: let `useTransition` own the mount/unmount lifecycle:**

```jsx
import { useTransition, animated } from '@react-spring/web';

function ExpandableItem({ items, openIndex }) {
  // Pass the currently-open item's content as the transition item
  const activeItem = items[openIndex] ?? null;

  const transitions = useTransition(activeItem, {
    key: activeItem?.id,  // change key = treat as new item = triggers leave on old
    from: { opacity: 0, height: 0, overflow: 'hidden' },
    enter: { opacity: 1, height: 'auto' },
    leave: { opacity: 0, height: 0 },
    config: { tension: 200, friction: 25 },
  });

  return transitions((style, item) =>
    item ? (
      <animated.div style={style}>
        {item.content}
      </animated.div>
    ) : null
  );
}
```

**The critical difference:** `useTransition` renders the `leave` animation on the **old** value simultaneously with the `enter` animation on the **new** value. React doesn't unmount the old element — react-spring keeps it in the DOM until `leave` finishes, then removes it.

**If you need `exitBeforeEnter` behavior** (old item fully exits before new one enters):
```js
exitBeforeEnter: true,
```

**Common mistake** causing the bug: wrapping `useTransition` output in a conditional:
```jsx
// WRONG — React unmounts before leave plays
{isOpen && transitions((style, item) => <animated.div style={style}>{item}</animated.div>)}

// CORRECT — let transitions handle mount/unmount
{transitions((style, item) => item && <animated.div style={style}>{item}</animated.div>)}
```
